Add a condition on the duplicated event names

mattiamonari opened the pull request:
### What does this PR do?
Add a condition on the name of the events of the java wrapper

### Where should the reviewer start?
There is just a one file.

### Why is it needed?
Based on Issue #1781, for contracts which have '**overloaded**' events (same name, but different signatures) the Java Wrapper creates those events as different instances of the Event class, but with the same variable name, thus resulting in an error. This PR adds the `changeEventsName()` method which uses an HashMap to count the occurences of an event name, and for those with more than one occurence, it will change the name putting a number at its end, e.g. 

```
public static final Event ACTIONPAUSED_EVENT = new Event("ActionPaused",
            Arrays.<TypeReference<?>>asList(new TypeReference<Utf8String>() {}, new TypeReference<Bool>() {}));
    ;

    public static final Event ACTIONPAUSED1_EVENT = new Event("ActionPaused", 
            Arrays.<TypeReference<?>>asList(new TypeReference<Address>() {}, new TypeReference<Utf8String>() {}, new TypeReference<Bool>() {}));
    ;
```

gtebrean commented on codegen/src/main/java/org/web3j/codegen/SolidityFunctionWrapper.java:
```diff
@@ -436,6 +437,25 @@ private List<MethodSpec> buildFunctionDefinitions(
         return methodSpecs;
     }
 
+    private void changeEventsName(List<AbiDefinition> functionDefinitions) {
```
```suggestion
    private void changeDuplicateEventsName(List<AbiDefinition> functionDefinitions) {
```

gtebrean commented on codegen/src/main/java/org/web3j/codegen/SolidityFunctionWrapper.java:
```diff
@@ -436,6 +437,25 @@ private List<MethodSpec> buildFunctionDefinitions(
+
+        Map<String, Integer> countMap = new HashMap<>();
+
+        for (AbiDefinition functionDefinition : functionDefinitions) {
+            if (TYPE_EVENT.equals(functionDefinition.getType())
+                    && functionDefinition.getName() != null) {
+                String s = functionDefinition.getName();
+                if (countMap.containsKey(s)) {
+                    int count = countMap.get(s);
+                    functionDefinition.setName(s + count);
+                    countMap.put(s, count + 1);
+                } else {
+                    countMap.put(s, 1);
+                }
+            }
+        }
```
```suggestion
        functionDefinitions.stream().filter(function -> TYPE_EVENT.equals(function.getType())
                && function.getName() != null).forEach(function -> {
            String functionName = function.getName();
            if (countMap.containsKey(functionName)) {
                int count = countMap.get(functionName);
                function.setName(functionName + count);
                countMap.put(functionName, count + 1);
            } else {
                countMap.put(functionName, 1);
            }
        });
```

gtebrean reviewed (commented):
This is a good workaround for now. Please add also a unit test to can include it in today release.

mattiamonari commented:
Added the unit test. Not sure if that's the proper way of doing it, please check. I had to put the buildFunctionDefinitions package-private since it was not accessible inside tests.

I also removed the usage of changeEventsName() in buildFuncNameConstants(), l.742 since the function operate just on TYPE_FUNCTION objects (correct me if I'm wrong).

mattiamonari commented:
Apparently my Unit Test is not working. On my local machine, anyway, it correctly builds and passes tests.

gtebrean commented on codegen/src/test/java/org/web3j/codegen/SolidityFunctionWrapperTest.java:
```diff
@@ -902,6 +902,133 @@ public void testBuildFuncNameConstants() throws Exception {
+                        + "      java.util.Arrays.<org.web3j.abi.TypeReference<?>>asList(new org.web3j.abi.TypeReference<org.web3j.abi.datatypes.Utf8String>() {}, new org.web3j.abi.TypeReference<org.web3j.abi.datatypes.Bool>() {}));\n"
+                        + "  ;\n"
+                        + "\n"
+                        + "  public static final org.web3j.abi.datatypes.Event EVENTNAME1_EVENT = new org.web3j.abi.datatypes.Event(\"eventName1\", \n"
```
Event name shouldn't be modified  `new org.web3j.abi.datatypes.Event(\"eventName1\"...` will fail in request.

mattiamonari replied:
Can you please give me some more details about that?
Do you mean it is not possible to use the abiDef.setName() function on TYPE_EVENTS objects?

Thanks

gtebrean commented:
I won't added it to this release as it affects the wrapper and has to be carefully tested also by us. 
We keep the PR open for finding a better solution.

gtebrean commented:
> @gtebrean can you please give me some more information if you have them? With the new commit the eventName passed in the constructor should be the same. In my local env. the tests passed, but apparently there's still something which I'm doing wrongly since the checks failed.

jobs are complaining that:
org.web3j.codegen.SolidityFunctionWrapperTest > testBuildFunctionDuplicatedEventNames() FAILED
    org.opentest4j.AssertionFailedError at SolidityFunctionWrapperTest.java:1038.
    
It fails to create the wrapper, make sure you have latest changes from master.

mattiamonari commented:
> > @gtebrean can you please give me some more information if you have them? With the new commit the eventName passed in the constructor should be the same. In my local env. the tests passed, but apparently there's still something which I'm doing wrongly since the checks failed.
> 
> jobs are complaining that: org.web3j.codegen.SolidityFunctionWrapperTest > testBuildFunctionDuplicatedEventNames() FAILED org.opentest4j.AssertionFailedError at SolidityFunctionWrapperTest.java:1038.
> 
> It fails to create the wrapper, make sure you have latest changes from master.

My branch when I commited was up-to-date with master. org.web3j.codegen.SolidityFunctionWrapperTest > testBuildFunctionDuplicatedEventNames() is the unit test I created my self for this PR, but in my local env is passing the test, but not in the jobs. Which could be the cause? Error logs aren't really useful.

SergioIOB commented:
Hi! @mattiamonari @gtebrean 

I would like to know the status of this, as we have recently encountered this problem as well.

Best regards

gtebrean commented:
#2124 fixed this, thank you @mattiamonari for taking the initiative and @SergioIOB for fixing the pipeline
